### PR TITLE
Bug 1173758 - [toolkit][ServiceWorkers] SimpleOfflineCache call CacheStorage.open should not be called in every fetch request

### PR DESCRIPTION
**Pull request by arcturus**
@lodr did the change to cache the promise call for getting the cache name.

Also did a major change to avoid copying the fetchAndCache provided by the chaceHelper library, so default policy for simpleoffline is fetch :S ... please tell me if you dont like it.

Also found another major problem, will add another bug is that chrome canary is not supporting postMessage through the sw clients, so the mocha adapter never return results or finish.

Using normal chrome works. I'll send a patch for using chrome stable, not cause of this, but cause the library works with Chrome stable, but we should keep an eye on what is goig to happen with postMessage

### CODE REVIEW

**Comment by arcturus** [comment]
btw, this fixes #22

File: lib/spec/simpleofflinecache.sw-spec.js [hunk 1]
```diff
@@ -85,19 +74,16 @@ describe('SimpleOfflineCache instance', function () {
...
-      return caches.open('test');
-    })
-    .then(function (cache) {
-      return cache.match(request.clone());
-    })
-    .then(function (response) {
-      expect(response).to.be.ok;
+      expect(fetch.calledOnce).to.be.true;
+      expect(fetch.getCall(0).args[0].url).to.equal(request.url);
+      fetch.restore();
```
Comment (lodr) [inline lib/spec/simpleofflinecache.sw-spec.js]: You should add this in the `catch()` handler as well to allow other tests to continue their execution independently.

File: lib/spec/simpleofflinecache.sw-spec.js [hunk 2]
```diff
@@ -118,16 +104,17 @@ describe('SimpleOfflineCache instance', function () {
     var request = new Request('/index.html?param=zzz');
 
     var cacheObj = {
-      match: sinon.stub().returns(Promise.resolve(null)),
-      put: sinon.stub().returns(Promise.resolve())
+      match: sinon.stub().returns(Promise.resolve(null))
     };
-    sinon.stub(self.cacheHelper, 'getCache').returns(
+    // Clean the cached request
+    delete offline.cacheRequest;
```
Comment (lodr) [inline lib/spec/simpleofflinecache.sw-spec.js]: It's very strange you need to delete the request from a new instance of `SimpleOfflineCache`. I would prefer you to assert about `offline.cacheRequest` to be `undefined` just after creating the instance rather than modifing it. Moreover, this is an internal so I would not touch it at all.

File: lib/spec/simpleofflinecache.sw-spec.js [hunk 3]
```diff
@@ -118,16 +104,17 @@ describe('SimpleOfflineCache instance', function () {
...
+    sinon.stub(caches, 'open').returns(
       Promise.resolve(cacheObj));
-    offline.onFetch(request).then(function() {
-      expect(self.cacheHelper.getCache.calledOnce).to.be.true;
+    offline.onFetch(request.clone()).then(function() {
```
Comment (lodr) [inline lib/spec/simpleofflinecache.sw-spec.js]: Why did you add the `.clone()` here?

File: lib/spec/simpleofflinecache.sw-spec.js [hunk 4]
```diff
@@ -118,16 +104,17 @@ describe('SimpleOfflineCache instance', function () {
...
+      expect(caches.open.calledOnce).to.be.true;
       expect(cacheObj.match.called).to.be.true;
       expect(cacheObj.match.calledWith(request, options)).to.be.true;
-      self.cacheHelper.getCache.restore();
+      caches.open.restore();
```
Comment (lodr) [inline lib/spec/simpleofflinecache.sw-spec.js]: Add to the `catch()` handler as well.

**Comment by lodr** [comment]
Indeed, @arcturus , I'm quite happy about missing policy to be just `fetch`, the simpler the better. That's the philosophy fo all sww.

I see you added some assertions about caching the offline cache to a preexistent test described as `should pass match parameters`. I think this should go in another different test in charge of simply test the `open` is called only once upon two different store operations.

File: lib/spec/simpleofflinecache.sw-spec.js (hunk 3 above)
Comment (arcturus) [inline lib/spec/simpleofflinecache.sw-spec.js]: I had to modify the mock for the cache.match, to return something, if i return null (as i was doing before), we will be doing a fetch, and will be consuming the request body, so comparing them was incorrect.

But you are right, the clone is not necessary there, just modify the `cache.match` mock.

File: lib/simpleofflinecache.js [hunk 5]
```diff
@@ -48,7 +48,7 @@ SimpleOfflineCache.prototype.onFetch = function soc_onFetch(request, response) {
   var _this = this;
   debug('Handing fetch event: %s', clone.url);
```
Comment (lodr) [inline lib/simpleofflinecache.js]: nit: let's get rid of that `%s`

File: package.json [hunk 6]
```diff
@@ -27,8 +27,5 @@
     "vinyl-buffer": "^1.0.0",
     "karma-firefox-launcher": "^0.1.4",
     "vinyl-source-stream": "^1.0.0"
-  },
```
Comment (lodr) [inline package.json]: nice

File: lib/spec/simpleofflinecache.sw-spec.js [hunk 7]
```diff
@@ -118,19 +105,21 @@ describe('SimpleOfflineCache instance', function () {
...
+      match: sinon.stub().returns(Promise.resolve({}))
     };
-    sinon.stub(self.cacheHelper, 'getCache').returns(
+    // Clean the cached request
+    delete offline.cacheRequest;
+    sinon.stub(caches, 'open').returns(
       Promise.resolve(cacheObj));
     offline.onFetch(request).then(function() {
-      expect(self.cacheHelper.getCache.calledOnce).to.be.true;
+      expect(caches.open.calledOnce).to.be.true;
```
Comment (lodr) [inline lib/spec/simpleofflinecache.sw-spec.js]: As stated in the [last general comment](https://github.com/gaia-components/serviceworkerware/pull/29#issuecomment-119858117) you should separate the test of caching the cache request in a different thread. Notice there you should call `onFetch()` and assert `caches.open` is called once.

File: lib/simpleofflinecache.js (hunk 5 above)
Comment (arcturus) [inline lib/simpleofflinecache.js]: yup
